app/invocation/invocation_build_logs_card.tsx

```diff
@@ -1,14 +1,137 @@
 }
 
-export default class BuildLogsCardComponent extends React.Component {
-  props: Props;
+interface State {
+  consoleBuffer?: string;
+  nextChunkId?: string;
+  fetchedFirstChunk?: boolean;
+}
+
+const POLL_TAIL_INTERVAL_MS = 1_000;
+const MAX_INITIAL_LINES = 100_000;
+
+const InvocationStatus = invocation.Invocation.InvocationStatus;
+
+export default class BuildLogsCardComponent extends React.Component<Props, State> {
+  state: State = {
+    consoleBuffer: "",
+    nextChunkId: "",
+    fetchedFirstChunk: false,
+  };
+
+  private pollTailTimeout: number | null = null;
+
+  constructor(props: Props) {
+    super(props);
+
+    const invocation = props.model.invocations[0];
+    if (!invocation) {
+      console.error("BuildLogsCard: invocation model is missing invocation");
+    }
+  }
+
+  componentDidMount() {
+    if (this.shouldFetchFromEventLog()) {
+      this.fetchTail(/*isFirstRequest=*/ true);
+    }
+  }
+
+  componentWillUnmount() {
+    window.clearTimeout(this.pollTailTimeout);
+  }
+
+  private shouldFetchFromEventLog() {
+    return capabilities.chunkedEventLogs && Boolean(this.props.model.invocations[0]?.lastChunkId);
+  }
+
+  private isInvocationComplete() {
+    const invocation = this.props.model.invocations[0];
+    return invocation?.invocationStatus === InvocationStatus.COMPLETE_INVOCATION_STATUS;
+  }
+
+  private fetchTail(isFirstRequest = false) {
+    const invocation = this.props.model.invocations[0];
+    let rpcError: BuildBuddyError | null = null;
+    let nextChunkId = "";
+    const wasCompleteBeforeMakingRequest = this.isInvocationComplete();
+
+    rpcService.service
+      .getEventLogChunk(
+        new eventlog.GetEventLogChunkRequest({
+          invocationId: invocation.invocationId,
+          chunkId: this.state.nextChunkId,
+          ...(isFirstRequest && {
+            // For the first request, fetch a large amount of lines starting from
+            // the tail and reading backward.
+            minLines: MAX_INITIAL_LINES,
+            readBackward: true,
+          }),
+        })
+      )
+      .then((response) => {
+        console.log(response);
+        let consoleBuffer = this.state.consoleBuffer;
+        for (const line of response.chunk?.lines || []) {
+          consoleBuffer += String.fromCharCode(...line) + "\n";
+        }
+        nextChunkId = response.nextChunkId;
+        this.setState({ consoleBuffer, nextChunkId });
+      })
+      .catch((e) => {
+        rpcError = BuildBuddyError.parse(e);
+      })
+      .finally(() => {
+        // NotFound / OutOfRange errors just mean the next chunk has not yet been written
+        // and that we should continue polling.
+        if (rpcError?.code === "NotFound" || rpcError?.code === "OutOfRange") {
+          // If the tail chunk of a completed invocation was not found, no new
+          // chunks should be written, so stop polling.
+          // Note: this relies on the server not writing any new chunks after an
```

Comment:
Should we just have the server return an empty `nextChunkId` when the whole log has been loaded? I feel like that simplifies a ton of this logic.

Reply:
IIUC it's because the chunkstore isn't aware of the invocation status, so it always has its `next_chunk_id` pointer set to the place where the next chunk would be written, regardless of whether it'd get written or not. We may actually be doing an invocation lookup for auth purposes anyway -- if so then maybe we can read the invocation status on the server and return empty next chunk id if it's completed -- @tempoz any thoughts?